lib/src/main/java/graphql/nadel/engine/transform/partition/NadelPartitionTransform.kt

```diff
@@ -0,0 +1,266 @@
+                executionBlueprint,
+                services,
+                service,
+                overallField,
+                hydrationDetails,
+            )
+
+        if (fieldPartitionContext == null) {
+            // A field without a partition context can't be partitioned
+            return null
+        }
+
+        val fieldPartition = NadelFieldPartition(
+            pathToPartitionArg = getPathToPartitionArg(overallField, executionBlueprint.engineSchema) ?: return null,
+            fieldPartitionContext = fieldPartitionContext,
+            graphQLSchema = executionBlueprint.engineSchema,
+            partitionKeyExtractor = partitionTransformHook.getPartitionKeyExtractor()
+        )
+
+        val fieldPartitions = try {
+            fieldPartition.createFieldPartitions(field = overallField)
+        } catch (exception: Exception) {
+            return State(
+                executionContext = executionContext,
+                fieldPartitionContext = fieldPartitionContext,
+                errors = mutableListOf(exception)
+            )
+        }
+
+        if (fieldPartitions.size < 2) {
+            // We can't partition a field that doesn't have at least two partitions
+            return null
+        }
+
+        partitionTransformHook.willPartitionCallback(executionContext, fieldPartitions)
+
+        return State(
+            executionContext = executionContext,
+            fieldPartitionContext = fieldPartitionContext,
+            fieldPartitions = fieldPartitions
+        )
+    }
+
+    private fun getPathToPartitionArg(overallField: ExecutableNormalizedField, schema: GraphQLSchema): List<String>?  {
+        val fieldDef = overallField.getFieldDefinitions(schema).first()
+        val routingDirective = fieldDef.getAppliedDirective(NadelDirectives.partitionDirectiveDefinition.name) ?: return null
+        val pathToSplitPoint = routingDirective.getArgument("pathToSplitPoint")
+            ?.argumentValue
+            ?.value as ArrayValue?
+
+        return pathToSplitPoint?.values?.map { it as StringValue }?.map { it.value }
+    }
+
+    override suspend fun transformField(
+        executionContext: NadelExecutionContext,
+        serviceExecutionContext: NadelServiceExecutionContext,
+        transformer: NadelQueryTransformer,
+        executionBlueprint: NadelOverallExecutionBlueprint,
+        service: Service,
+        field: ExecutableNormalizedField,
+        state: State,
+    ): NadelTransformFieldResult {
+
+        if (state.errors != null) {
+            // At this point we know we can't partition the field, but removing it could result
+            // in a query with no top-level fields, which would be invalid.
+            // Adding __typename as an artificial field also wouldn't work because that field
+            // belongs to all operation types, which causes issues further down the execution.
+            return NadelTransformFieldResult.unmodified(field)
```

Comment:
I feel like we should just stop the execution here?

Like could perhaps we just `throw SomeNadelException` or so and it should include that in the response if the thrown type is a `GraphQLError` I think.